### DISCUSSION TITLE
makeObservable nested annotations

### DISCUSSION THREAD

**Discussion by bellostom** [mobxjs/mobx#3260] [General]
Hi

Given the below object

```
{
    "name" : "Project",
    "status": "enabled",  -> observable
    "tasks": [{ -> observable.shallow
          "id" : 1,
           "status": "enabled"  -> observable
          "name": "Example",
          "exercises": [{    -> observable.shallow
               "code": 12,
               "name": "Test"   -> observable
          }]
     }]
}

```

Is there a way to pass annotations it to makeObservable so that it observes only the attributes I have marked above?

Thanks for any help

**Comment by urugator** [comment]
You can nest the `observable/makeObservable` calls. But if you eg. want to add new task, you have to make it observable manually before pushing into the array.
```javascript
makeObservable({
    "name" : "Project",
    "status": "enabled",
    "tasks": [makeObservable({
          "id" : 1,
           "status": "enabled"
          "name": "Example",
          "exercises": [{
               "code": 12,
               "name": "Test"
          }, { status: observable, exercises: observable.shallow })]
     }]
}, { status: observable, tasks: observable.shallow })
```

**Reply by bellostom** [reply]
Hi @urugator 
Thanks for the reply.
I wonder how much overhead will be added if I just make the whole object an observable, considering that the number of objects will be anywhere from 3000 to 4000 entries

**Reply by urugator** [reply]
My guess is, if making everything observable would be slow for your use case, then excluding few things here and there, won't make it faster enought to satisfy the use case. But as always when it comes to optimization you just have to try.
Foremost I would ask myself: Do I need to load thousands of objects in the same moment? Does the user need to see them all immediately? What's he gonna do with them?

**Reply by bellostom** [reply]
These objects are going to be lazy loaded in the background and they will not be at any point required to be visible all at the same time.
However, they will be in memory for the entire time the application is accessible which is why I am looking into how much overhead making each object a whole observable, would incur. 
I will run some tests with memory profiling and different scenarios

**Reply by urugator** [reply]
I see, memory wise, there could be quite a difference, but with 4000 objects we are probably talking about units of megabytes.
Please share your findings and make sure to test it with production build.

**Comment by bellostom** [comment]
Hi @urugator 
Found some time to do some testing

Test scenario is this
a list of 220 objects. Each JSON object is 140kb. Each object contains a list of 380 tasks

Results
Storing the list in memory as decoded by JSON.parse consumes 103mb
Storing the list in memory as decoded by JSON.parse and { status: observable, exercises: observable.shallow } consumes 425mb
Storing the list in memory as decoded by JSON.parse and passing each object by observable consumes 595mb

As it seems the overhead is considerable

In order to keep memory usage low, I intend to make each object observable once they need to be and remove observability once they are no longer need to be (but keep them stored in memory).

What is the proposed way of removing observables from an object

Regards

**Reply by urugator** [reply]
> Test scenario is this

Ah, so it's 83600, not 4000 :) Once you get to tens/hundreds of MBs it becomes quite bad indeed.

> observables from an object

Do you mean "observability"? I guess no other way than a deep copy, for simple objects `toJS` should suffice.
Btw may I ask what's the idea behind keeping these in memory? Should it work as a cache, avoiding requests?

**Reply by bellostom** [reply]
@urugator 
Well it maybe even more than 83600 :)
This is just a sample from some more accurate data I could gather.
You are right regarding the need to have them in memory. It is kind of a cache indeed.
Will check toJS since I will need to keep observability in a "per need" basis
